### ISSUE TITLE
Thaw failure with serialized functions - seeking workaround

### ISSUE THREAD

**Issue by hsartoris-bard** [juji-io/datalevin#92]
Version: 0.5.28

Upfront: I have no expectation that this project fully support the following use case; just seeing if there are any reasonable workarounds for the situation.

### Background
I'm using [nippy-serializable-fns](https://github.com/redplanetlabs/nippy-serializable-fns) to allow serializing fns into the database, which in my case is being used more as a runtime cache/log with the benefits of datalog querying and persistence across application restarts.

Unsurprisingly, the application restarts are proving to be the challenge with frozen functions. I can transact new entities with similar functions into the database and retrieve them without issue, but entities transacted prior to a restart produce the following:
```
Execution error (ExceptionInfo) at datalevin.scan/get-value (scan.cljc:145).
Fail to get-value: "Thaw failed: Decryption/decompression failure, or data unfrozen/damaged."
```

<details>
  <summary>Full stacktrace</summary>

  ```clojure
#error {
 :cause "Fail to get-value: \"Thaw failed: Decryption/decompression failure, or data unfrozen/damaged.\""
 :data {:dbi "datalevin/giants", :k 222, :k-type :id, :v-type :datom}
 :via
 [{:type clojure.lang.ExceptionInfo
   :message "Fail to get-value: \"Thaw failed: Decryption/decompression failure, or data unfrozen/damaged.\""
   :data {:dbi "datalevin/giants", :k 222, :k-type :id, :v-type :datom}
   :at [datalevin.scan$get_value invokeStatic "scan.cljc" 145]}]
 :trace
 [[datalevin.scan$get_value invokeStatic "scan.cljc" 145]
  [datalevin.scan$get_value invoke "scan.cljc" 137]
  [datalevin.binding.java.LMDB get_value "java.clj" 395]
  [datalevin.binding.java.LMDB get_value "java.clj" 393]
  [datalevin.storage$retrieved__GT_datom invokeStatic "storage.cljc" 166]
  [datalevin.storage$retrieved__GT_datom invoke "storage.cljc" 161]
  [datalevin.storage.Store head "storage.cljc" 354]
  [datalevin.db.DB _first "db.cljc" 107]
  [datalevin.db$transact_add invokeStatic "db.cljc" 686]
  [datalevin.db$transact_add invoke "db.cljc" 668]
  [datalevin.db$local_transact_tx_data$fn__52940 invoke "db.cljc" 883]
  [datalevin.db$local_transact_tx_data invokeStatic "db.cljc" 733]
  [datalevin.db$local_transact_tx_data invoke "db.cljc" 731]
  [datalevin.db$transact_tx_data invokeStatic "db.cljc" 967]
  [datalevin.db$transact_tx_data invoke "db.cljc" 950]
  [datalevin.core$with invokeStatic "core.cljc" 259]
  [datalevin.core$with invoke "core.cljc" 254]
  [datalevin.core$_transact_BANG_$fn__55825 invoke "core.cljc" 479]
  [clojure.lang.Atom swap "Atom.java" 37]
  [clojure.core$swap_BANG_ invokeStatic "core.clj" 2356]
  [clojure.core$swap_BANG_ invoke "core.clj" 2349]
  [datalevin.core$_transact_BANG_ invokeStatic "core.cljc" 478]
  [datalevin.core$_transact_BANG_ invoke "core.cljc" 475]
  [datalevin.core$transact_BANG_ invokeStatic "core.cljc" 573]
  [datalevin.core$transact_BANG_ invoke "core.cljc" 485]
  [datalevin.core$transact_BANG_ invokeStatic "core.cljc" 570]
  [datalevin.core$transact_BANG_ invoke "core.cljc" 485]
  [user$eval119878 invokeStatic "NO_SOURCE_FILE" 1]
  [user$eval119878 invoke "NO_SOURCE_FILE" 1]
  [clojure.lang.Compiler eval "Compiler.java" 7181]
  [clojure.lang.Compiler eval "Compiler.java" 7136]
  [clojure.core$eval invokeStatic "core.clj" 3202]
  [clojure.core$eval invoke "core.clj" 3198]
  [nrepl.middleware.interruptible_eval$evaluate$fn__107569$fn__107570 invoke "interruptible_eval.clj" 87]
  [clojure.lang.AFn applyToHelper "AFn.java" 152]
  [clojure.lang.AFn applyTo "AFn.java" 144]
  [clojure.core$apply invokeStatic "core.clj" 667]
  [clojure.core$with_bindings_STAR_ invokeStatic "core.clj" 1977]
  [clojure.core$with_bindings_STAR_ doInvoke "core.clj" 1977]
  [clojure.lang.RestFn invoke "RestFn.java" 425]
  [nrepl.middleware.interruptible_eval$evaluate$fn__107569 invoke "interruptible_eval.clj" 87]
  [clojure.main$repl$read_eval_print__9110$fn__9113 invoke "main.clj" 437]
  [clojure.main$repl$read_eval_print__9110 invoke "main.clj" 437]
  [clojure.main$repl$fn__9119 invoke "main.clj" 458]
  [clojure.main$repl invokeStatic "main.clj" 458]
  [clojure.main$repl doInvoke "main.clj" 368]
  [clojure.lang.RestFn invoke "RestFn.java" 1523]
  [nrepl.middleware.interruptible_eval$evaluate invokeStatic "interruptible_eval.clj" 84]
  [nrepl.middleware.interruptible_eval$evaluate invoke "interruptible_eval.clj" 56]
  [nrepl.middleware.interruptible_eval$interruptible_eval$fn__107600$fn__107604 invoke "interruptible_eval.clj" 152]
  [clojure.lang.AFn run "AFn.java" 22]
  [nrepl.middleware.session$session_exec$main_loop__107668$fn__107672 invoke "session.clj" 202]
  [nrepl.middleware.session$session_exec$main_loop__107668 invoke "session.clj" 201]
  [clojure.lang.AFn run "AFn.java" 22]
  [java.lang.Thread run "Thread.java" 829]]}
  ```
</details>

I can pull attributes for these entities other than those containing functions without issue.

### Ideal solution
I have the requisite data to rebuild the impacted (single-valued) attributes and replace their values, and this would be easy enough to do on application startup. However, because transacting in a new datom involves [retrieving](https://github.com/juji-io/datalevin/blob/eebf22d76e5ed038468cb9dfb8d9533fc03865c2/src/datalevin/db.cljc#L683) any current value, this does not work. Because of the various indices that are maintained, I'm hesitant about attempting to manually strip the offending values from the dbi.

**TL;DR:** is there any way to clobber or evict an existing datom without requiring that its value be parseable? Or do I need to just make sure that I remove any such values before application shutdown?

**Comment by den1k** [comment]
@hsartoris-bard can you provide a minimal code example to reproduce the issue?

**Comment by huahaiy** [comment]
As @den1k indicated, maybe some example code of what you were doing is helpful to fully understand the case.

My guess is that if you store the serialized fns as bytes and declare them as such in your schema, it should work.

As to your current problem of removing broken datoms, it is possible to export the data as text file, manually edit them out, then import the data.

**Comment by hsartoris-bard** [comment]
I'm finding it tricky to get a reliable repro working. It turns out not to have been quite all of the previously transacted functions; possibly something to do with those containing closures. My suspicion is that I was unintentionally abusing nippy-serializable-fns, and that the functions genuinely were not deserializable due to changes in compiled classes or the like. I have gotten it to occur sporadically with the below example by running it repeatedly and making more or less random changes to the functions within.
<details>
  <summary>datalevin_test.clj</summary>

  ```clojure
(ns datalevin-test
  (:require
    [com.rpl.nippy-serializable-fn]
    [datalevin.core :as d]))

(def closed-over-var 3)

(defn f [x y] (+ closed-over-var x y))

(defn make-blob [z]
  (let [s (into [] (take 1000) (repeatedly #(rand-int 10)))
        ssum (apply + s)]
    [{:a [(fn [x] (f x z))]
      :big-thing s
      :sum ssum
      :b [{:pred (fn [x] (f x ssum))}]}]))

(defn something-else [x y z]
  (f x (+ x y z)))

(defn -main [& args]
  (let [conn (d/get-conn "test-db")]
    (something-else 1 2 3)
    (if (= "query" (first args))
      ;; it will fail here, if anywhere
      (println
        (d/q '[:find (pull ?e [*]) :where (or [?e ::f] [?e ::blob])] @conn))
      (println
        (d/transact! conn [[:db/add -1 ::f f]
                           [:db/add -2 ::blob (make-blob 10)]])))
    (d/close conn)))
```
</details>

Broadly, however, I have a couple of attributes that store EDN blobs that contain functions, usually multiple functions per blob. The attributes are implicitly single-value, but are not declared in the schema precisely because they really are big EDN blobs. The structure of one might look like:
```clojure
[{:phase-1 [{:pred some?, :path [:a :b]}
            ...]
  :phase-2 [{:pred (fn [x] (some-util-fn x expected-value))}
            ...]}
 {:phase-1 ...}]
```

They are being serialized implicitly by the extensions to Nippy made by requiring `com.rpl.nippy-serializable-fn`.

I think the approach suggested by @huahaiy is probably most informative in this situation. Because I have the key into datalevin/giants for a given problematic entity, I could use `dtlv -d <db-dir> -a dump` to dump the raw data, then process the tuples belonging to datalevin/giants to map from raw->ids, e.g.:
```clojure
(-> "AAAAAAAAAAE"
    datalevin.util/decode-base64
    ByteBuffer/wrap
    #'datalevin.bits/get-long)
#_=> 1
```
Use that table to locate the encoded key value for the `long` key, then strip any tuples from any DBI that contain that string in the dump; rinse and repeat.

I feel like there must be an easier way to do this programmatically, but probably the best approach is not to get into this situation in the first place.
